### PR TITLE
New package `happy-lib`; make previous `happy-*` packages sublibraries (#288)

### LINKED ISSUE #288: Should we make the `happy-*` packages public library components of `happy`?
Cabal 3.0 introduced the [`library:visibility` field](https://cabal.readthedocs.io/en/3.4/cabal-package.html#pkg-field-library-visibility).
Should we use it to define the single-library-component `happy-*` packages as "sublibraries"? Depending on such a library is exemplified here: https://github.com/haskell/cabal/issues/9480#issuecomment-1826883582. For us it would lead to names such as `happy:tabular`.
Defining such public sublibraries is exemplified here: https://github.com/BlockScope/plugins-for-blobs/blob/develop/plugins-for-blobs.cabal. Note the `visibility: public` fields.

Pros:

* No separate .cabal file, change log, package description, supported versions, etc. for each package to maintain. I think this is huge, because maintainer time is scarce.

Cons:
 
* Needs Cabal 3.0 (released in 2019). 

Given that

1. we already decided to lock into the GHC ecosystem 
2. `cabal-install` is easily buildable with older GHCs
3. older `happy` versions will continue to build with ancient GHCs and accept the same .y-file syntax,

I don't think requiring Cabal 3.0 is much of a drawback. On the other hand I *do* see the real costs associated with release management if we stick to one executable package and 5 library packages, all of which need to keep their versions in sync etc.

**This decision has potential to lock us in for quite a long time, as `happy-tabular` is a library name (technically a package name which implicitly denotes the default library) than `happy:tabular`. I would like to have the Go from @int-index, @Ericson2314 and @andreasabel on this.**

### PR DESCRIPTION
**Pull request by sgraf812**
This is one design design discussed in #288.

Fixes #288.

### CODE REVIEW

File: lib/happy-lib.cabal [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        Happy.Frontend.ParseMonad.Class
+        Happy.Frontend.Mangler.Monad
+        Happy.Frontend.Parser
+        Happy.Frontend.Lexer
+        Happy.Frontend.ParamRules
+        Happy.Frontend.AttrGrammar
+        Happy.Frontend.AttrGrammar.Parser
+        Happy.Frontend.AttrGrammar.Mangler
+
+library tabular
```
Comment (int-index) [inline lib/happy-lib.cabal:102]: What's the relation between this `library tabular` subsection and `lib/tabular/happy-tabular.cabal`? I didn't expect to see the latter.

Comment (Ericson2314) [reply lib/happy-lib.cabal:102]: I think we can just remove the others?

**Review by Ericson2314 — COMMENTED** [review]
Same question as @int-index about old cabal files, but very big +1 on the design!

File: lib/happy-lib.cabal (hunk 1 above)
Comment (sgraf812) [reply lib/happy-lib.cabal:102]: Good point, I forgot to remove the .cabal files. The main thing about this approach :)

Should be done now.

**Review by int-index — APPROVED** [review]

**Comment by sgraf812** [comment]
Alright then, let's do it.

**Pull request merged**